File: TolometCore/src/com/akrog/tolomet/Measurement.java
package com.akrog.tolomet;

import java.util.LinkedHashMap;
import java.util.Map;

public class Measurement {
	private final Map<Long, Number> map = new LinkedHashMap<Long, Number>();
	private Long[] times;
	private Number[] values;
	
	private void clearCache() {
		times = null;
		values = null;
	}
	
	public void put(long time, Number value ) {
		map.put(time, value);
		clearCache();
	}
	
	public Long[] getTimes() {
		if( times == null )
			times = map.keySet().toArray(new Long[0]); 
		return times;
	}
	
	public Long getStamp() {
		if( isEmpty() )
			return null;
		return getTimes()[size()-1];
	}
	
	public Number getLast() {
		if( isEmpty() )
			return null;
		return getValues()[size()-1];
	}
	
	public Number[] getValues() {
		if( values == null )
			values = map.values().toArray(new Number[0]); 
		return values;
	}
	
	public int size() {
		return map.size();
	}
	
	public void clear() {
		map.clear();
		clearCache();
	}
	
	public void clear( long fromStamp ) {
		for( Long stamp : getTimes() )
			if( stamp < fromStamp )
				map.remove(stamp);
		clearCache();
	}
	
	public boolean isEmpty() {
		return map.isEmpty();
	}
	
	public void merge(Measurement data) {
		map.putAll(data.map);
		clearCache();
	}
}

File: Project/tolometShared/src/commonMain/kotlin/com/akrog/tolomet/Measurement.kt
package com.akrog.tolomet

class Measurement(
    private val validMinimum: Float? = null,
    private val validMaximum: Float? = null
) {
    private val map = mutableMapOf<Long, Number>()
    private var times: Array<Long>? = null
    private var values: Array<Number>? = null
    private var cachedMinimum: Number? = null
    private var cachedMaximum: Number? = null

    private fun clearCache() {
        times = null
        values = null
        cachedMinimum = null
        cachedMaximum = null
    }

    fun put(time: Long, value: Number?) {
        if (value == null) return
        if (validMinimum != null && value.toFloat() < validMinimum) return
        if (validMaximum != null && value.toFloat() > validMaximum) return

        // Note: Simple rounding to the minute as in the original Java code
        val roundedTime = (time / 60000) * 60000

        map[roundedTime] = value
        clearCache()
    }

    fun getEntrySet(): Set<Map.Entry<Long, Number>> = map.entries

    fun getTimes(): Array<Long> {
        if (times == null) {
            times = map.keys.sorted().toTypedArray()
        }
        return times!!
    }

    fun getValues(): Array<Number> {
        if (values == null) {
            val sortedKeys = getTimes()
            values = sortedKeys.map { map[it]!! }.toTypedArray()
        }
        return values!!
    }

    fun isEmpty(): Boolean = map.isEmpty()

    fun getStamp(): Long? {
        if (isEmpty()) return null
        val times = getTimes()
        return times.last()
    }

    fun getBegin(): Long? {
        if (isEmpty()) return null
        val times = getTimes()
        return times.first()
    }

    fun get(time: Long): Number? = map[time]

    fun clear() {
        map.clear()
        clearCache()
    }

    fun clear(fromStamp: Long) {
        val toRemove = map.keys.filter { it >= fromStamp }
        toRemove.forEach { map.remove(it) }
        clearCache()
    }

    fun merge(other: Measurement) {
        map.putAll(other.map)
        clearCache()
    }
}
